engine/runtime-compiler/src/main/java/org/enso/compiler/pass/lint/unusedimports/UnusedImports.java

```diff
@@ -0,0 +1,217 @@
+          diff.removeAll(usedSymbolsForImp);
+          if (!diff.isEmpty()) {
+            var warn = createWarning(imp, diff);
+            imp.getDiagnostics().add(warn);
+          }
+        } else {
+          var usedSymbolsForImp = usedSymbols.getUsedSymbolsForImport(imp);
+          if (usedSymbolsForImp.isEmpty()) {
+            var warn = createWarning(imp);
+            imp.getDiagnostics().add(warn);
+          }
+        }
+      }
+    }
+    return ir;
+  }
+
+  /** Returns set of all imported symbol by the given import statement. */
+  private Set<QualifiedName> importedSymbols(Import impIr, BindingsMap bindingsMap) {
+    var resolvedImp = findResolvedImport(impIr, bindingsMap);
+    if (resolvedImp == null) {
+      var errMsg = new StringBuilder();
+      errMsg
+          .append("No resolved import found for import statement '")
+          .append(impIr.showCode())
+          .append("' in module '")
+          .append(bindingsMap.currentModule().getName())
+          .append("'");
+      var loc = impIr.identifiedLocation();
+      if (loc != null) {
+        errMsg.append("at location: {").append(loc).append("}");
+      }
+      errMsg.append(".");
+      throw new CompilerError(errMsg.toString());
+    }
+    var symbols = importedSymbols(resolvedImp);
+    return new HashSet<>(symbols);
+  }
+
+  private static List<QualifiedName> importedSymbols(BindingsMap.ResolvedImport resolvedImport) {
+    var impDef = resolvedImport.importDef();
+    if (impDef.onlyNames().isDefined()) {
+      var targets = resolvedImport.targets();
+      if (targets.size() != 1) {
+        throw new AssertionError(
+            "Resolved import for '"
+                + impDef.showCode()
+                + "' should have a single target."
+                + " Instead, targets are: "
+                + resolvedImport.targets());
+      }
+      var target = targets.head();
+      var names = impDef.onlyNames().get().map(Literal::name);
+      var resolvedNames = new ArrayList<QualifiedName>();
+      names.foreach(
+          name -> {
+            var expSymbols = target.findExportedSymbolsFor(name);
+            expSymbols.foreach(
+                expSymbol -> {
+                  resolvedNames.add(expSymbol.qualifiedName());
+                  return null;
+                });
+            return null;
+          });
+      return resolvedNames;
+    } else {
+      var names = resolvedImport.targets().map(ResolvedName::qualifiedName);
+      return CollectionConverters.asJava(names);
+    }
+  }
+
+  private static BindingsMap.ResolvedImport findResolvedImport(
+      Import impIr, BindingsMap bindingsMap) {
+    for (var resolvedImp : CollectionConverters.asJava(bindingsMap.resolvedImports())) {
+      if (resolvedImp.importDef() == impIr || haveSameLocations(resolvedImp.importDef(), impIr)) {
+        return resolvedImp;
+      }
+    }
+    return null;
+  }
+
+  private static boolean haveSameLocations(Import imp1, Import imp2) {
```

Comment:
First incarnation of the same logic.

Reply:
Using single method in e44d0ccf558bb50c40bd081b8db1a92165383998